File: packages/py/llm_core/llm_core/providers/openai_compatible.py
"""OpenAI-compatible provider.

KEY LESSON: OpenAI's HTTP API became a de-facto standard. OpenAI itself,
NVIDIA NIM, Ollama, vLLM, LM Studio, Together, Groq, and many others all expose
the SAME ``/v1/chat/completions`` and ``/v1/embeddings`` shape. So one class —
just a different ``base_url`` + ``api_key`` + model id — covers three of our
four providers. Only Anthropic ships its own distinct API.
"""

from __future__ import annotations

from collections.abc import Iterable, Iterator

from openai import OpenAI

from ..types import (
    ChatMessage,
    ChatOptions,
    ChatResult,
    EmbeddingResult,
    TokenUsage,
    to_messages,
)


class OpenAICompatibleProvider:
    def __init__(
        self,
        name: str,
        api_key: str,
        chat_model: str,
        embed_model: str,
        base_url: str | None = None,
    ) -> None:
        self.name = name
        self.chat_model = chat_model
        self.embed_model = embed_model
        self._client = OpenAI(api_key=api_key, base_url=base_url)

    def _payload(self, messages: Iterable[ChatMessage | dict], options: ChatOptions | None):
        opts = options or ChatOptions()
        msgs = [{"role": m.role, "content": m.content} for m in to_messages(messages)]
        kwargs: dict = {"model": opts.model or self.chat_model, "messages": msgs}
        if opts.temperature is not None:
            kwargs["temperature"] = opts.temperature
        if opts.max_tokens is not None:
            kwargs["max_tokens"] = opts.max_tokens
        if opts.stop:
            kwargs["stop"] = opts.stop
        return kwargs

    def chat(
        self,
        messages: Iterable[ChatMessage | dict[str, str]],
        options: ChatOptions | None = None,
    ) -> ChatResult:
        kwargs = self._payload(messages, options)
        resp = self._client.chat.completions.create(**kwargs)
        usage = resp.usage
        return ChatResult(
            text=resp.choices[0].message.content or "",
            model=kwargs["model"],
            usage=TokenUsage(
                input_tokens=getattr(usage, "prompt_tokens", None),
                output_tokens=getattr(usage, "completion_tokens", None),
            ),
            raw=resp,
        )

    def chat_stream(
        self,
        messages: Iterable[ChatMessage | dict[str, str]],
        options: ChatOptions | None = None,
    ) -> Iterator[str]:
        kwargs = self._payload(messages, options)
        kwargs["stream"] = True
        for chunk in self._client.chat.completions.create(**kwargs):
            # Some OpenAI-compatible servers (LM Studio, vLLM, …) emit chunks with
            # an empty `choices` list — e.g. a final usage-only chunk. Skip them.
            if not chunk.choices:
                continue
            delta = chunk.choices[0].delta.content
            if delta:
                yield delta

    def embed(self, input: list[str]) -> EmbeddingResult:
        resp = self._client.embeddings.create(model=self.embed_model, input=input)
        return EmbeddingResult(
            vectors=[d.embedding for d in resp.data],
            model=self.embed_model,
            usage=TokenUsage(input_tokens=getattr(resp.usage, "prompt_tokens", None)),
        )

File: packages/ts/llm-core/src/providers/openai-compatible.ts
/**
 * OpenAI-compatible provider.
 *
 * KEY LESSON: OpenAI's HTTP API became a de-facto standard. OpenAI itself,
 * NVIDIA NIM, Ollama, vLLM, LM Studio, Together, Groq, and many others all
 * expose the SAME `/v1/chat/completions` and `/v1/embeddings` shape. So one
 * class — just a different `baseURL` + `apiKey` + model id — covers five of
 * our six providers. Only Anthropic ships its own distinct API.
 */

import OpenAI from "openai";
import type {
  ChatMessage,
  ChatOptions,
  ChatResult,
  EmbeddingResult,
  LLMProvider,
} from "../types.js";

export interface OpenAICompatConfig {
  name: string;
  apiKey: string;
  baseURL?: string;
  chatModel: string;
  embedModel: string;
}

export class OpenAICompatibleProvider implements LLMProvider {
  readonly name: string;
  readonly chatModel: string;
  readonly embedModel: string;
  private client: OpenAI;

  constructor(cfg: OpenAICompatConfig) {
    this.name = cfg.name;
    this.chatModel = cfg.chatModel;
    this.embedModel = cfg.embedModel;
    this.client = new OpenAI({ apiKey: cfg.apiKey, baseURL: cfg.baseURL });
  }

  async chat(messages: ChatMessage[], options: ChatOptions = {}): Promise<ChatResult> {
    const model = options.model ?? this.chatModel;
    const resp = await this.client.chat.completions.create({
      model,
      messages,
      temperature: options.temperature,
      max_tokens: options.maxTokens,
      stop: options.stop,
    });
    return {
      text: resp.choices[0]?.message?.content ?? "",
      model,
      usage: {
        inputTokens: resp.usage?.prompt_tokens,
        outputTokens: resp.usage?.completion_tokens,
      },
      raw: resp,
    };
  }

  async *chatStream(
    messages: ChatMessage[],
    options: ChatOptions = {},
  ): AsyncIterable<string> {
    const model = options.model ?? this.chatModel;
    const stream = await this.client.chat.completions.create({
      model,
      messages,
      temperature: options.temperature,
      max_tokens: options.maxTokens,
      stop: options.stop,
      stream: true,
    });
    for await (const chunk of stream) {
      const delta = chunk.choices[0]?.delta?.content;
      if (delta) yield delta;
    }
  }

  async embed(input: string[]): Promise<EmbeddingResult> {
    const resp = await this.client.embeddings.create({
      model: this.embedModel,
      input,
      // Force plain-float output. With no encoding_format the OpenAI SDK
      // requests base64 and decodes it client-side; some OpenAI-compatible
      // servers (e.g. LM Studio) don't honour that, yielding all-zero vectors.
      encoding_format: "float",
    });
    return {
      vectors: resp.data.map((d) => d.embedding),
      model: this.embedModel,
      usage: { inputTokens: resp.usage?.prompt_tokens },
    };
  }
}
